File: src/lib/scanner/manifest.ts
import type { AppInfo } from "./types";
import { getAllStrings } from "./extract";

function findStringValue(strings: string[], key: string): string | undefined {
  const idx = strings.findIndex((s) => s === key || s.includes(key));
  if (idx === -1) return undefined;

  for (let i = idx + 1; i < Math.min(idx + 5, strings.length); i++) {
    const val = strings[i];
    if (val && !val.startsWith("android.") && val.length > 1 && val.length < 200) {
      return val;
    }
  }
  return undefined;
}

function extractPermissions(strings: string[]): string[] {
  return [...new Set(strings.filter((s) => s.startsWith("android.permission.")))];
}

function extractComponents(strings: string[], marker: string): string[] {
  const components: string[] = [];
  for (let i = 0; i < strings.length; i++) {
    if (strings[i] === marker || strings[i].endsWith(marker)) {
      const name = strings[i + 1];
      if (name && name.includes(".") && !name.startsWith("android.")) {
        components.push(name);
      }
    }
  }
  return [...new Set(components)];
}

function parseManifestBoolean(strings: string[], attr: string): boolean | undefined {
  const idx = strings.findIndex((s) => s === attr || s.includes(attr));
  if (idx === -1) return undefined;

  for (let i = idx; i < Math.min(idx + 4, strings.length); i++) {
    if (strings[i] === "true") return true;
    if (strings[i] === "false") return false;
  }
  return undefined;
}

export function parseManifest(
  manifestData: Uint8Array | undefined,
  fileName: string,
  fileSize: number,
): AppInfo {
  const strings = manifestData ? getAllStrings(manifestData) : [];

  const packageName =
    findStringValue(strings, "package") ??
    strings.find((s) => /^[a-z][a-z0-9_]*(\.[a-z][a-z0-9_]*)+$/.test(s) && s.length > 5);

  return {
    fileName,
    fileSize,
    packageName,
    versionName: findStringValue(strings, "versionName"),
    versionCode: findStringValue(strings, "versionCode"),
    minSdk: findStringValue(strings, "minSdkVersion"),
    targetSdk: findStringValue(strings, "targetSdkVersion"),
    permissions: extractPermissions(strings),
    activities: extractComponents(strings, "activity"),
    services: extractComponents(strings, "service"),
    receivers: extractComponents(strings, "receiver"),
    providers: extractComponents(strings, "provider"),
    debuggable: parseManifestBoolean(strings, "debuggable"),
    allowBackup: parseManifestBoolean(strings, "allowBackup"),
    usesCleartextTraffic: parseManifestBoolean(strings, "usesCleartextTraffic"),
  };
}

export function parseNetworkSecurityConfig(data: Uint8Array | undefined): {
  cleartextPermitted: boolean;
  hasPinning: boolean;
} {
  if (!data) return { cleartextPermitted: false, hasPinning: false };
  const text = new TextDecoder("utf-8", { fatal: false }).decode(data);
  return {
    cleartextPermitted: /cleartextTrafficPermitted\s*=\s*"true"/i.test(text),
    hasPinning: /<pin-set/i.test(text),
  };
}

File: scanner/internal/manifest/manifest.go
package manifest

import (
	"bufio"
	"os"
	"regexp"
	"strings"

	"github.com/apkbugfinder/scanner/internal/grep"
	"github.com/apkbugfinder/scanner/internal/types"
)

var (
	rePackage      = regexp.MustCompile(`package="([^"]+)"`)
	reVersionName  = regexp.MustCompile(`versionName="([^"]+)"`)
	reVersionCode  = regexp.MustCompile(`versionCode="([^"]+)"`)
	reMinSDK       = regexp.MustCompile(`minSdkVersion="([^"]+)"`)
	reTargetSDK    = regexp.MustCompile(`targetSdkVersion="([^"]+)"`)
	rePermission   = regexp.MustCompile(`<uses-permission[^>]+android:name="([^"]+)"`)
	reActivity     = regexp.MustCompile(`<activity[^>]+android:name="([^"]+)"`)
	reService      = regexp.MustCompile(`<service[^>]+android:name="([^"]+)"`)
	reReceiver     = regexp.MustCompile(`<receiver[^>]+android:name="([^"]+)"`)
	reProvider     = regexp.MustCompile(`<provider[^>]+android:name="([^"]+)"`)
	reAllowBackup  = regexp.MustCompile(`android:allowBackup="true"`)
	reDebuggable   = regexp.MustCompile(`android:debuggable="true"`)
	reCleartext    = regexp.MustCompile(`android:usesCleartextTraffic="true"`)
	reExportedTrue = regexp.MustCompile(`android:exported="true"`)
	reNetSecConf   = regexp.MustCompile(`android:networkSecurityConfig="@xml/([^"]+)"`)
)

func Parse(manifestPath string, fileName string, fileSize int64) (types.AppInfo, string, error) {
	info := types.AppInfo{
		FileName:    fileName,
		FileSize:    fileSize,
		Permissions: []string{},
		Activities:  []string{},
		Services:    []string{},
		Receivers:   []string{},
		Providers:   []string{},
	}

	data, err := os.ReadFile(manifestPath)
	if err != nil {
		return info, "", err
	}
	content := string(data)

	if m := rePackage.FindStringSubmatch(content); len(m) > 1 {
		info.PackageName = m[1]
	}
	if m := reVersionName.FindStringSubmatch(content); len(m) > 1 {
		info.VersionName = m[1]
	}
	if m := reVersionCode.FindStringSubmatch(content); len(m) > 1 {
		info.VersionCode = m[1]
	}
	if m := reMinSDK.FindStringSubmatch(content); len(m) > 1 {
		info.MinSDK = m[1]
	}
	if m := reTargetSDK.FindStringSubmatch(content); len(m) > 1 {
		info.TargetSDK = m[1]
	}

	info.Permissions = unique(rePermission.FindAllStringSubmatch(content, -1))
	info.Activities = unique(reActivity.FindAllStringSubmatch(content, -1))
	info.Services = unique(reService.FindAllStringSubmatch(content, -1))
	info.Receivers = unique(reReceiver.FindAllStringSubmatch(content, -1))
	info.Providers = unique(reProvider.FindAllStringSubmatch(content, -1))

	if reAllowBackup.MatchString(content) {
		t := true
		info.AllowBackup = &t
	}
	if reDebuggable.MatchString(content) {
		t := true
		info.Debuggable = &t
	}
	if reCleartext.MatchString(content) {
		t := true
		info.UsesCleartextTraffic = &t
	}

	info.ComponentSummary = ComponentSummary(content)

	netConf := ""
	if m := reNetSecConf.FindStringSubmatch(content); len(m) > 1 {
		netConf = m[1]
	}

	return info, netConf, nil
}

func unique(matches [][]string) []string {
	seen := map[string]bool{}
	var out []string
	for _, m := range matches {
		if len(m) > 1 && !seen[m[1]] {
			seen[m[1]] = true
			out = append(out, m[1])
		}
	}
	return out
}

func ComponentSummary(content string) types.ComponentSummary {
	return types.ComponentSummary{
		ExportedActivities: countExportedBlocks(content, "<activity"),
		ExportedProviders:  countExportedBlocks(content, "<provider"),
		ExportedReceivers:  countExportedBlocks(content, "<receiver"),
		ExportedServices:   countExportedBlocks(content, "<service"),
	}
}

func countExportedBlocks(content, tag string) int {
	count := 0
	idx := 0
	for {
		start := strings.Index(content[idx:], tag)
		if start == -1 {
			break
		}
		start += idx
		end := strings.Index(content[start:], ">")
		if end == -1 {
			break
		}
		block := content[start : start+end]
		if strings.Contains(block, `android:exported="true"`) {
			count++
		}
		idx = start + end
	}
	return count
}

func ExportedWithoutPermission(manifestPath string) ([]grep.Match, error) {
	f, err := os.Open(manifestPath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var matches []grep.Match
	scanner := bufio.NewScanner(f)
	lineNum := 0
	for scanner.Scan() {
		lineNum++
		line := scanner.Text()
		if !strings.Contains(line, `<service`) &&
			!strings.Contains(line, `<activity`) &&
			!strings.Contains(line, `<provider`) &&
			!strings.Contains(line, `<receiver`) {
			continue
		}
		if strings.Contains(line, `android:exported="true"`) && !strings.Contains(line, `android:permission="`) {
			matches = append(matches, grep.Match{
				File:    manifestPath,
				Line:    lineNum,
				Content: strings.TrimSpace(line),
			})
		}
	}
	return matches, scanner.Err()
}

func GrepManifest(manifestPath string, opts grep.Options) ([]grep.Match, error) {
	return grep.SearchFile(manifestPath, opts)
}
